src/adc.rs

```diff
@@ -0,0 +1,858 @@
+//! Analog-to-digital converter
+//!
+use core::marker::PhantomData;
+
+use crate::delay::McycleDelay;
+use crate::gpio::gpioa::{PA0, PA1, PA2, PA3, PA4, PA5, PA6, PA7};
+use crate::gpio::gpiob::{PB0, PB1};
+use crate::gpio::gpioc::{PC0, PC1, PC2, PC3, PC4, PC5};
+use crate::gpio::Analog;
+use crate::hal::adc::{Channel, OneShot};
+use crate::pac::{ADC0, ADC1};
+use crate::rcu::{Enable, Rcu, Reset};
+
+macro_rules! adc_pins {
+    ($ADC:ident, $($input:ty => $chan:expr),+ $(,)*) => {
+        $(
+            impl Channel<$ADC> for $input {
+                type ID = u8;
+
+                fn channel() -> u8 {
+                    $chan
+                }
+            }
+        )+
+    };
+}
+
+/// Contains types related to ADC configuration
+#[allow(non_camel_case_types)]
+pub mod config {
+
+    /// The number of cycles to sample a given channel for
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    #[allow(non_camel_case_types)]
+    #[repr(u8)]
+    pub enum SampleTime {
+        /// 1.5 cycles
+        Point_1_5 = 0,
+        /// 7.5 cycles
+        Point_7_5,
+        /// 13.5 cycles
+        Point_13_5,
+        /// 28.5 cycles
+        Point_28_5,
+        /// 41.5 cycles
+        Point_41_5,
+        /// 55.5 cycles
+        Point_55_5,
+        /// 71.5 cycles
+        Point_71_5,
+        /// 239.5 cycles
+        Point_239_5,
+    }
+
+    impl Default for SampleTime {
+        fn default() -> Self {
+            SampleTime::Point_55_5
+        }
+    }
+
+    /// Clock config for the ADC
+    /// Check the datasheet for the maximum speed the ADC supports
+    #[derive(Debug, Clone, Copy)]
+    pub enum Clock {
+        /// ADC prescaler select CK_APB2/2
+        Apb2_div_2 = 0,
+        /// ADC prescaler select CK_APB2/4
+        Apb2_div_4 = 1,
+        /// ADC prescaler select CK_APB2/6
+        Apb2_div_6 = 2,
+        /// ADC prescaler select CK_APB2/8
+        Apb2_div_8 = 3,
+        /// ADC prescaler select CK_APB2/12
+        Apb2_div_12 = 5,
+        /// ADC prescaler select CK_APB2/16
+        Apb2_div_16 = 7,
+    }
+
+    /// Resolution to sample at
+    #[derive(Debug, Clone, Copy)]
+    #[repr(u8)]
+    pub enum Resolution {
+        /// 12-bit ADC resolution
+        Twelve = 0,
```

Comment:
To be coherent with other enums, maybe something like `Res_12, … Res_6` would be preferable.

Reply:
https://docs.rs/stm32f4xx-hal/0.8.3/stm32f4xx_hal/adc/config/enum.Resolution.html